packages/babel-compiler/babel-compiler.js

```diff
@@ -173,57 +173,64 @@ BCp._inferFromBabelRc = function (inputFile, babelOptions, cacheDeps) {
   if (babelrcPath) {
     if (! hasOwn.call(this._babelrcCache, babelrcPath)) {
       try {
-        this._babelrcCache[babelrcPath] =
-          JSON.parse(inputFile.readAndWatchFile(babelrcPath));
+        this._babelrcCache[babelrcPath] = {
+          controlFilePath: babelrcPath,
+          controlFileData: JSON.parse(
+            inputFile.readAndWatchFile(babelrcPath)),
+          deps: Object.create(null),
+        };
       } catch (e) {
         if (e instanceof SyntaxError) {
           e.message = ".babelrc is not a valid JSON file: " + e.message;
         }
-
         throw e;
       }
     }
 
-    return this._inferHelper(
-      inputFile,
-      babelOptions,
-      babelrcPath,
-      this._babelrcCache[babelrcPath],
-      cacheDeps
-    );
+    const cacheEntry = this._babelrcCache[babelrcPath];
+
+    if (this._inferHelper(inputFile, cacheEntry)) {
+      merge(babelOptions, cacheEntry, "presets");
+      merge(babelOptions, cacheEntry, "plugins");
+      Object.assign(cacheDeps, cacheEntry.deps);
+      return true;
+    }
   }
 };
 
 BCp._inferFromPackageJson = function (inputFile, babelOptions, cacheDeps) {
   var pkgJsonPath = inputFile.findControlFile("package.json");
   if (pkgJsonPath) {
-    if (! hasOwn.call(this._babelrcCache, pkgJsonPath)) {
-      this._babelrcCache[pkgJsonPath] = JSON.parse(
-        inputFile.readAndWatchFile(pkgJsonPath)
-      ).babel || null;
+    const cacheEntry = hasOwn.call(this._babelrcCache, pkgJsonPath)
+      ? this._babelrcCache[pkgJsonPath]
+      : this._babelrcCache[pkgJsonPath] = {
+          controlFilePath: pkgJsonPath,
+          controlFileData: JSON.parse(
+            inputFile.readAndWatchFile(pkgJsonPath)
+          ).babel || null,
+          deps: Object.create(null),
+        };
+
+    if (this._inferHelper(inputFile, cacheEntry)) {
+      merge(babelOptions, cacheEntry, "presets");
+      merge(babelOptions, cacheEntry, "plugins");
+      Object.assign(cacheDeps, cacheEntry.deps);
+      return true;
     }
-
-    return this._inferHelper(
-      inputFile,
-      babelOptions,
-      pkgJsonPath,
-      this._babelrcCache[pkgJsonPath],
-      cacheDeps
-    );
   }
 };
 
-BCp._inferHelper = function (
-  inputFile,
-  babelOptions,
-  controlFilePath,
-  babelrc,
-  cacheDeps
-) {
-  if (! babelrc) {
+BCp._inferHelper = function (inputFile, cacheEntry) {
+  if (! cacheEntry.controlFileData) {
     return false;
   }
 
+  if (hasOwn.call(cacheEntry, "finalInferHelperResult")) {
+    // We've already run _inferHelper and populated
+    // cacheEntry.{presets,plugins}, so we can return early here.
+    return cacheEntry.finalInferHelperResult;
+  }
```

Comment:
@veered I think this early return may help with the unnecessary `.babelrc` processing that you noticed using [`qualia:profile`](https://github.com/qualialabs/profile).

Reply:
Woo!! Thanks for looking into it :)

Reply:
Thanks!